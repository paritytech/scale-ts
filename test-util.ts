/// <reference lib="deno.unstable"/>

import { assertEquals, assertThrows } from "https://deno.land/std@0.161.0/testing/asserts.ts"
import { assertSnapshot } from "https://deno.land/std@0.161.0/testing/snapshot.ts"
import { AnyShape, assert, DecodeBuffer, Shape, ShapeAssertError } from "./common/mod.ts"

const [lipsum, words, cargoLock] = ["lipsum.txt", "words.txt", "Cargo.lock"].map((fileName) => () =>
  Deno.readTextFile(fileName)
)
export const files = { lipsum: lipsum!, words: words!, cargoLock: cargoLock! }

export function testShape<T>(
  shape: Shape<T>,
  values: NoInfer<T>[] | Record<string, NoInfer<T> | (() => NoInfer<T>)>,
  async?: boolean,
): void
export function testShape<T>(
  shape: Shape<T>,
  values: T[] | Record<string, T | (() => T)>,
  async = false,
) {
  for (const key in values) {
    let value = values[key as never] as T | (() => T)
    const label = values instanceof Array
      ? Deno.inspect(value, {
        depth: Infinity,
        trailingComma: true,
        compact: true,
        iterableLimit: Infinity,
      })
      : key
    Deno.test(`${Deno.inspect(shape)} ${label}`, async (t) => {
      if (typeof value === "function") {
        value = (value as () => T)()
      }
      assert(shape, value)
      const encoded = async ? await shape.encodeAsync(value) : shape.encode(value)
      await assertSnapshot(t, encoded, { serializer: serializeU8A })
      const decodeBuffer = new DecodeBuffer(encoded)
      const decoded = shape.subDecode(decodeBuffer)
      assertEquals(decoded, value)
      assertEquals(decodeBuffer.index, encoded.length)
      assert(shape, decoded)
    })
  }
}

export function testInvalid(shape: AnyShape, values: unknown[] | Record<string, unknown | (() => unknown)>) {
  for (const key in values) {
    let value: unknown = values[key as never]
    const label = values instanceof Array
      ? Deno.inspect(value, {
        depth: Infinity,
        trailingComma: true,
        compact: true,
        iterableLimit: Infinity,
      })
      : key
    Deno.test(`${Deno.inspect(shape)} invalid ${label}`, async (t) => {
      if (typeof value === "function") {
        value = (value as () => unknown)()
      }
      await assertThrowsSnapshot(t, () => assert(shape as Shape<any>, value), ShapeAssertError)
    })
  }
}

function serializeU8A(array: Uint8Array) {
  return [...array].map((x) => x.toString(16).padStart(2, "0")).join("\n")
}

type NoInfer<T> = T extends infer U ? U : never
export function benchShape<T>(name: string, shape: Shape<T>, value: NoInfer<T>): void
export function benchShape<T>(name: string, shape: Shape<T>, value: T) {
  const encoded = shape.encode(value)
  Deno.bench(`- ${name} (${encoded.length}B) [encode] `, () => {
    shape.encode(value)
  })
  Deno.bench(`  ${name} (${encoded.length}B) [decode]`, () => {
    shape.decode(encoded)
  })
}

export async function assertThrowsSnapshot(
  t: Deno.TestContext,
  fn: () => unknown,
  ctor: abstract new(...args: any) => Error = Error,
) {
  try {
    fn()
  } catch (error) {
    if (!(error instanceof ctor)) {
      throw new Error("Expected " + ctor.name)
    }
    return await assertSnapshot(t, `${error.name}: ${error.message}`, { serializer: (x) => x })
  }
  throw new Error("Expected function to throw")
}

export { assertEquals, assertSnapshot, assertThrows }
