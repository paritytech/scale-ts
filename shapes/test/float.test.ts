import * as $ from "../../mod.ts"
import { testInvalid, testShape } from "../../test-util.ts"

testShape($.f64, [
  0,
  1,
  -1,
  1.2345,
  1e23,
  4e-56,
  Number.MAX_SAFE_INTEGER,
  Number.MIN_SAFE_INTEGER,
  Number.MAX_VALUE,
  -Number.MAX_VALUE,
  NaN,
  -Infinity,
  Infinity,
])

testInvalid($.f64, ["0", null, {}])
