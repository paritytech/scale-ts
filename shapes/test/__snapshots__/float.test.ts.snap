export const snapshot = {};

snapshot[`\$.f64 0 1`] = `
00
00
00
00
00
00
00
00
`;

snapshot[`\$.f64 1 1`] = `
00
00
00
00
00
00
f0
3f
`;

snapshot[`\$.f64 -1 1`] = `
00
00
00
00
00
00
f0
bf
`;

snapshot[`\$.f64 1.2345 1`] = `
8d
97
6e
12
83
c0
f3
3f
`;

snapshot[`\$.f64 1e+23 1`] = `
f6
4a
e1
c7
02
2d
b5
44
`;

snapshot[`\$.f64 4e-56 1`] = `
d2
c0
57
b2
b0
62
6f
34
`;

snapshot[`\$.f64 9007199254740991 1`] = `
ff
ff
ff
ff
ff
ff
3f
43
`;

snapshot[`\$.f64 -9007199254740991 1`] = `
ff
ff
ff
ff
ff
ff
3f
c3
`;

snapshot[`\$.f64 1.7976931348623157e+308 1`] = `
ff
ff
ff
ff
ff
ff
ef
7f
`;

snapshot[`\$.f64 -1.7976931348623157e+308 1`] = `
ff
ff
ff
ff
ff
ff
ef
ff
`;

snapshot[`\$.f64 NaN 1`] = `
00
00
00
00
00
00
f8
7f
`;

snapshot[`\$.f64 -Infinity 1`] = `
00
00
00
00
00
00
f0
ff
`;

snapshot[`\$.f64 Infinity 1`] = `
00
00
00
00
00
00
f0
7f
`;

snapshot[`\$.f64 invalid "0" 1`] = `ShapeAssertError: typeof value !== "number"`;

snapshot[`\$.f64 invalid null 1`] = `ShapeAssertError: typeof value !== "number"`;

snapshot[`\$.f64 invalid {} 1`] = `ShapeAssertError: typeof value !== "number"`;
